FrSky FLVSS Individual Cell Voltages Telemetry

AustinSaintAubin opened the pull request:
Added FrSky FLVSS Individual Cell Voltages "Cels" Telemetry Emulation to S.Port Telemetry.
Has accuracy of 0.01, tested to support up to 6 cells (might support more the 6 cells)

To view individual cells or more importantly to get lowest cell (all cells are the sum of vbat, so each cell is the same in this case):
See [OpenTX 2.1 & FrSky FLVSS Individual Cell Voltages](http://openrcforums.com/forum/viewtopic.php?t=7266).
> Add a new sensor.
> 
> To display the lowest cell voltage set it up like this
> Type: Calculated
> Formula: Cell
> Cell Sensor: Cels _(pack total voltage, sum of all cells)_
> Cell Index: Lowest

3djc commented:
I fail to see the value of that,  since it is kinda misleading value,  as you are not sending real cell value but just VFAS divided by cell count for each cell,  hence not adding information to the existing VFAS.

AustinSaintAubin commented:
Adding this S.Port information makes if very easy on OpenTX and other systems to get accurate Pack & Cell voltages at the addition of very little overhead. This allows you to swap 2, 3, or 4 cell packs on your aircraft without having to change the voltage alarms. This also give a more accurate representation of the packs voltage because of added precision.

See my code in use here(YouTube): [Cleanflight FrSky FLVSS Individual Cell Voltages Telemetry](https://youtu.be/tfvRa6ey_9I)

Also, this is how I have my alarms configured, (see youtube video for how I have telemetry configured).
![img_1221](https://cloud.githubusercontent.com/assets/969780/13549612/2e4ac50a-e2cf-11e5-985a-c5545c8bc7e2.JPG)
![img_1222](https://cloud.githubusercontent.com/assets/969780/13549613/322355b6-e2cf-11e5-98b5-90f531457184.JPG)

3djc commented:
I clearly understand the benefits of individual cell measurement and FLVSS sensors, as I'm using those in addition to CF telem on demanding aircraft (and some quad definitively stress the batteries) to detect a Lipo that is starting to degrade (ie one cell is going lower than the others) using the tools that Heisenberg and the nice folks at the french taranis forum have put together 

![image](https://cloud.githubusercontent.com/assets/5167938/13553018/b3c4d126-e379-11e5-9bec-a044d213cb0e.png)

but what you are proposing is not that at all, It is just sending VFAS/CellNumber for every cell.

Also, if you carry multiple lipo size, have cells send through your proposed way is quite dangerous because a very depleted/defective 4 cells lipo can be seen as a 3 cells and will report fully chartged cells when they are not at all

You can do smart things with VFAS already, so I personnaly do not see a need to send fake Cells data

![image](https://cloud.githubusercontent.com/assets/5167938/13553070/db3351cc-e37b-11e5-8855-2dc90f55fb0b.png)

AustinSaintAubin commented:
Very nice [LI-PO Battcheck Script](http://frskytaranis.forumactif.org/t2800-lua-download-un-testeur-de-batterie-sur-la-radio).
Is there currently an easy way to get the assumed cell voltage from VFAS?
I see your point, but until I added FLVSS for Cels & Cell I had a lot of trouble setting appropriate alerts when all I had to work with was pack voltage. I still think that as long as it's known that cleanflight FLVSS data is a division of the pack voltage it should be fine. Formating the vbat data into FLVSS just makes it easier to work with. 

I will add a settings flag so sending FLVSS can be disable on CF. This will allow you to chain the telem port with a real FLVSS sensor and get CF telem + REAL cell value on the radio.
This will allow the use of a real FLVSS sensor, or a simulated one from CF.

AustinSaintAubin commented:
Added CLI setting so FLVSS can be disabled in the case where someone can use a real FLVSS sensor.

What do you think? Also, thanks for you input. I think having it so it can be disabled when a user wants the real thing is important.

3djc commented:
I was trying to find a better alternative when I encoutered this discussion (https://github.com/opentx/opentx/issues/2054) where it sounds like the opentx team does not like either sending fake cells instead of vfas.

So if I get your idea correctly, you are not used to work with total pack voltage, but feel confortable with a cell value? Fair enough, that's a personal thing I guess, buut it probably should be treated at radio level. There seems to be an issue with openttx and vfas ratio that I'll take with them.

You can fairly easyly create a custom mixer scripts that would do the math for you on your radio, something like : 

`-- LUA Mix Script
-- Return the average calculated value of a single cell based on vfas (vfas/cell number)
-- The data is returned is mv x 10 (ie 3.3 volts give 33)
-- This does not represent real cell value, just a calculated average

local outputs = { "1cel" }

local function run()

	cells_cnt = math.ceil(getValue("VFAS") / 4.25)
	if cells_cnt > 0 then
		return (getValue("VFAS")  * 100 / cells_cnt)
	end

end

return { run=run, output=outputs }`

then you set your alarms based on 1cel value

3djc commented:
After discussion with OpenTX team mebers, it appears the upcomming version of OpenTx will include changes to "ratio" which will do what you want to achieve, you'll just ratio your FVAS

AustinSaintAubin commented:
I updated my pull request to use `telemetry_flvss_cells` to enable and disable both Frsky & S.Port Cells output.

I tried to test your script after I [updated my firmware to include LUA custom mixer scripts](https://dsbeach.gitbooks.io/opentx-lua-reference-guide/content/getting_started.html), but I could not get the script to work, just reported error.
![screen-2016-03-07-162956 copy](https://cloud.githubusercontent.com/assets/969780/13585584/393e6e94-e482-11e5-8f43-028262919262.jpg)

I did however write a [test script](https://dsbeach.gitbooks.io/opentx-lua-reference-guide/content/output_table_syntax.html) based on one found in the [OpenTX 2.1 Lua Reference Guide](https://www.gitbook.com/book/dsbeach/opentx-lua-reference-guide/details). I used this to test how I could work with the outputs. The problem with the custom mixer scripts is that they are not usable on the telemetry screen, or at least I was not able to pull one up on the telemetry screen... This means that I can't do things like putting assumed cell voltage on the top bar.

The "ratio" solution looks like a better option than LUA custom mixes scripts.
Can't wait to test "ratio".
Please post here when better ratio comes to OpenTX.

For the time being I still think that having CF send FLVSS cells data to make working with the telemetry data regarding bat voltage easier for some, like me. You have brought up some very very good points.
I changed it so that *FLVSS is disabled by default.
Users will have to run `set telemetry_flvss_cells = ON` to enable FLVSS / Cels.
This so that people will have to read some documentation (like this post) to know the dangers of assumed cell voltages. Until "ratio" is out and tested, I think my solution still works fine.

3djc commented:
Yes, lua mixers have still a lot limits, maybe because they are too powerfull for a radio :dancer: I'll let you know when the ratio changes are out (and there is a MAJOR change to FC telemetry (not sensors telemetry) comming down the line as well, but not before 2.2 I'm hearing).

I thought I would quote one of the the opentx team member on this : "If you want cell reporting put an FLVSS on board, it's neither that big or expensive and you gain safety for your packs."

I guess many people do not realise they can just do that : chain CF telemetry and frsky s.port sensors, I probably need to write something in the docs about that.

Thanks a lot for listening to my "data integrist" rant !

AustinSaintAubin commented:
Thanks for your input. I think you're right about using a real FLVSS and all the other hazards of using an assumed cell value. But I still think I am right in this one case, with this being the easiest way for a person to monitor cell value for flight with all the assumed risk at this time.

Also, one of my problems is that I can't fit a real FLVSS in some of my planes.
In addition, with a lot of people trying to go under 0.55lb, every bit of weight counts.

I hope my code gets used so that people have the **option** to send assumed cells values. Having the setting disabled by default forces people to make a choice.

ledvinap commented on src/main/telemetry/smartport.c:
```diff
@@ -328,6 +328,35 @@ void handleSmartPortTelemetry(void)
+                    
+                    // Cells Data Payload
+                    uint32_t payload = 0;
+                    payload |= ((uint16_t)(vbat * 100 + batteryCellCount) / (batteryCellCount * 2)) & 0x0FFF;  // Cell Voltage formatted for payload, modified code from cleanflight Frsky.c, TESTING NOTE: (uint16_t)(4.2 * 500.0) & 0x0FFF;
```
2 cell values are sent in single packet, 12bit each

see:
https://github.com/opentx/opentx/blob/master/radio/src/telemetry/frsky_sport.cpp#L197

here is my implementation (rounding of cellVoltage is missing)
https://github.com/ledvinap/cleanflight/blob/local/src/main/telemetry/sport.c#L279

BTW: shifting values into correct position (as in frsky sendVoltage) is much easier to read ...

ledvinap commented:
The conditional flvss_cells is IMO a bit non-systematic. Is there anyone who needs it? What if I connect sensor that supplies another telemetry value?
Multiple IDs are reserved for each telemetry value and OpenTX seems to handle duplicate sensors (?). So maybe just move to different ID? Or move the conflicting sensor?
This may be handled in future PR.

CELLS message contain information about cell count detected by CF. This value can't be reliably calculated on TX. Is there other way to send this info?

Maybe OpenTX should define new telemetry messages that will match data available on flight controller. Then nobody will try to re-purpose existing ones ...

3djc commented:
There are not a lot of  sensors that conflict with CF values, or that you would want to overide FC data, because they are usualy accurate. On cells that would be very different tho, not everyone is having a racing drone and what to save 6 gramms, I'd rather save my lipos than 6 grams.

Instead of sending x time a similar value and risk conflic with a rather ligitimate external sensors, why don't we send once (per cycle obviously) the calculated average value using A3 or A4, which are PREC2 voltage sensors ?

3djc commented:
I've implemented sendiing the value through A3, which I think respond to Austin needs (can use for alarm, display on top bar,...). Precision is 0.01 and refresh rate is very fast. You can have a look ;

https://github.com/3djc/cleanflight/tree/cell-volt-on-a3

If both of you are ok with this, I'll write the corresponding doc change a issue a PR

AustinSaintAubin commented:
I think that [your solution of using A3](https://github.com/3djc/cleanflight/commit/6c5e7ee884e4a9f498e3a6a83dd83fa287a34f1f) is a better idea than FLVSS in the long term.
It's a lot less complicated to implement, and a lot smaller code size for the firmware.

I still need to test your code, but it looks good at first glance.
Might want to use A4 instead of A3 (to sperate it a little from A1 & A2 from receiver).
Also, do you think we need to have the ability to disable A3/A4 like my code has?

3djc commented:
@AustinSaintAubin As far as I know, a3/a4 are basicaly unused by sensors right now. I think they where added for an RPM/Voltage sensor, and do not seem to be used currently by others than DIY projects, so we should be fine and not disturbing anyone, even without the option to disable the code. If someone comes and says it breaks my 'unlikely' sensor, we know how to fix that :)

As for A3 or A4, it does not matter, they are the same specs, I just took the first in line, but if you think A4 makes more sense, I will make the change. Anyhow, this will need to go to the telemetry doc,

If you need me to compile a target for you, let me know !

3djc commented:
Move to A4 done, doc updated, ready to PR after your tests (https://github.com/3djc/cleanflight/tree/cell-volt-on-a3)

AustinSaintAubin commented:
Tested your [A4 code](https://github.com/3djc/cleanflight/tree/cell-volt-on-a3). Looks good.
The resolution of A4 is ~0.04. But that is the because of your code, just making a note of it.

hydra commented on src/main/telemetry/frsky.c:
```diff
@@ -358,37 +358,39 @@ static void sendVario(void)
+        payload |= ((cellVoltage & 0x0ff) << 8);
+
+        // Higher voltage bits are at bits 13-15
+        payload |= ((cellVoltage & 0xf00) >> 8);
```
to allow deletion of comments perhaps use a suitable `#define`s  The words `SHIFT` and `MASK` would likely be useful when coming up with names for the `#define`s.

ledvinap replied:
The format of packet is not too logical. Hiding the structure into #defines will make it even more complicated / it will be necessary to examine defines  along with the code

AustinSaintAubin replied:
@hydra I am not sure what you mean in your comment.
> to allow deletion of comments perhaps use a suitable #defines The words SHIFT and MASK would likely be useful when coming up with names for the #defines.

The code change here is just to encapsulate that code with:
if (telemetryConfig->telemetry_flvss_cells) { CODE }

hydra commented on src/main/config/config.c:
```diff
@@ -783,6 +783,7 @@ void validateAndFixConfig(void)
     // hardware supports serial port inversion, make users life easier for those that want to connect SBus RX's
 #ifdef TELEMETRY
     masterConfig.telemetryConfig.telemetry_inversion = 1;
+    //masterConfig.telemetryConfig.telemetry_flvss_cells = 1;  // ON/Enabled by default
```
delete commented out code please.

hydra replied:
delete commented out code please.

hydra replied:
delete commented out code please.

hydra commented on src/main/io/serial_cli.c:
```diff
@@ -544,6 +544,7 @@ const clivalue_t valueTable[] = {
 #ifdef TELEMETRY
     { "telemetry_switch",           VAR_UINT8  | MASTER_VALUE | MODE_LOOKUP,  &masterConfig.telemetryConfig.telemetry_switch, .config.lookup = { TABLE_OFF_ON } },
     { "telemetry_inversion",        VAR_UINT8  | MASTER_VALUE | MODE_LOOKUP,  &masterConfig.telemetryConfig.telemetry_inversion, .config.lookup = { TABLE_OFF_ON } },
+    { "telemetry_flvss_cells",      VAR_UINT8  | MASTER_VALUE | MODE_LOOKUP,  &masterConfig.telemetryConfig.telemetry_flvss_cells, .config.lookup = { TABLE_OFF_ON } },
```
not sure if the name of this is correct, since it only applies to frsky telemtry it should probably be something like 'frsky_send_cell_voltages`.

AustinSaintAubin replied:
"telemetry_flvss_cells" applies to both frsky & sport telemetry sent from CF.

3djc commented:
I think Austin would agree ,  the cell on a4 solution solves all the needs and is simpler, shouldn't we give it a try?

AustinSaintAubin commented:
I agree that @3djc solution is better long term because of proposed risk and additional steps that my solution requires. But, both (My's & @3djc's) solutions could be implemented to give users options. Would like some additional input from the dev community on the matter.

cjbeck commented:
I'm the one that opened the original bug on this: https://github.com/cleanflight/cleanflight/issues/1268.

I maintain that we're lacking functionality on the X4R receiver that the D4R receiver already gives us (the ability for the radio to do the per cell calculation). It seemed like a simple thing to do in the s.port code for Cleanflight (either pass the number of cells, or do the calc in cleanflight and pass as a different ID). Either way works for me. I just want to have a single model that I can swap batteries on and not have to change my telemetry alerts for.

Also, the argument about a single cell going bad and dragging the rest of the average will be a risk regardless of how you set up your telemetry. That's why you'd have to run the FrSky s.port lipo checker that plugs into your balance leads. But, those of us in the FPV quad racing community don't want to use that extra space / weight.

AustinSaintAubin commented:
What is the best way to resolve this branch pull request conflict... I just can't figure it out.

MasterZap commented:
Wait stop, this was all looking good until someone derailed it into A4? WTH? That defeates the whole purpouse of the feature!!!

The purpouse of the feature is this:

I have 10 copters all running Beta/Cleanflight. Some are 3s, some are 4s, one is 6s. I run them ALL with the same exact "model" in my Taranis (no switching on the radio necessary, just throw in an acc and fly)

On a few of them I have FLVSS sensors, with cell lever warnings set up.

On a few I don't, and use vBat.

The whole POINT of this feature, is to make it so that the SAME set of cell-level warnings can be used on ALL COPTERS. To do this, the data MUST be on the Cells value!!!

Why on earth this move to A4?? I don't get it!!!??

/Z

(Btw, I was just about to implement exactly this myself, because I needed it, because some of my copters are getting so small an FLVss would be unnecessary additional weight, luckily I found this PR, but got horrified as I read the comment and the A4 stuff!!)

AustinSaintAubin commented:
@MasterZap A4 is an option proposed by @3djc that is part of a separate pull request. 

My pull request adds the function you are looking for by emulating a FLVSS Sensor.
You can enable/disable this feature with "set telemetry_flvss_cells = ON/OFF".
Make sure that "telemetry_flvss_cells" is enabled on the quadcopters that you do not have a FLVSS Sensor on.

I am using my build on all my airframes and it works very well. Just waiting for it to get merged with main.

MasterZap commented:
ah ok.... sadly I´m on BetaFlight and have to wait until @borisbstyle merges this in.... it´s so funny I started looking at coding this EXACT thing today...... luckily I found your PR before wasting too much energy on it :)

AustinSaintAubin commented:
Ready to be pulled.

AustinSaintAubin commented:
What do I need to do to get this pulled?

hydra commented:
@AustinSaintAubin time from me, which I now have :smile: looking now...

hydra commented:
merging now, seems fine.

hydra commented:
@AustinSaintAubin after further review I made the following three commits.  Please review.

f1c7ecccaec3e425c614e811c290edce8500d407
af503135c3e4c7e9793876edd5641f850ff00261
848061e202effdd739b940f4110cada37f2f40c2
